Verbose, DefaultProgramArgs(), ...

Konstantin8105 opened the pull request:
Created for #232
Used external repository https://github.com/Konstantin8105/Cleaner

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/elliotchance/c2go/235)
<!-- Reviewable:end -->

elliotchance commented:
Reviewed 2 of 4 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks broke.

---

*[main.go, line 28 at r2](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KvteT07u8vEJ2MDIb6s:-KvteT07u8vEJ2MDIb6t:b-tdw8eh) ([raw file](https://github.com/elliotchance/c2go/blob/97383239879874a4b1cbed2d0098911930cd122d/main.go#L28)):*
> ```Go
> 	"errors"
> 
> 	cleaner "github.com/Konstantin8105/Cleaner"
> ```

I understand you want to maintain this package yourself, and you still can. However, this needs to be built in to `c2go` to be properly versioned with it.

---

*[main.go, line 220 at r2](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KvtfmieeoTmxo0KSmvt:-KvtfmieeoTmxo0KSmvu:bbl2u62) ([raw file](https://github.com/elliotchance/c2go/blob/97383239879874a4b1cbed2d0098911930cd122d/main.go#L220)):*
> ```Go
> 		err := cleaner.Go(outputFilePath, outputFilePath)
> 		if err != nil {
> 			fmt.Fprint(os.Stderr, "warning: Cannot removing unused variable, types, ....\nplease use flag '-keep-unused'")
> ```

You need to print out the actual error here so people can report it as a bug:

```go
fmt.Fprintf(os.Stderr, "warning: cannot removing unused entities: %s", err.Error())
```

---

*[main.go, line 233 at r2](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KvthNoG7IbzxZ1QDqaA:-KvthNoG7IbzxZ1QDqaB:b2xh4d9) ([raw file](https://github.com/elliotchance/c2go/blob/97383239879874a4b1cbed2d0098911930cd122d/main.go#L233)):*
> ```Go
> 	outputFlag        = transpileCommand.String("o", "", "output Go generated code to the specified file")
> 	packageFlag       = transpileCommand.String("p", "main", "set the name of the generated package")
> 	keepUnused        = transpileCommand.Bool("keep-unused", false, "Keep unused constants, functions, ...")
> ```

Please be more specific about what it being removed. Is it only constants and functions?

---

*[main_test.go, line 317 at r2](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kvti1QEPBWXP-ja94bx:-Kvti1QEPBWXP-ja94by:b5ni0n8) ([raw file](https://github.com/elliotchance/c2go/blob/97383239879874a4b1cbed2d0098911930cd122d/main_test.go#L317)):*
> ```Go
> 
> // newTempFile - returns temp file
> func newTempFile(dir, prefix, suffix string) (*os.File, error) {
> ```

Please use https://golang.org/pkg/io/ioutil/#TempFile instead.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/elliotchance/c2go/235)*
<!-- Sent from Reviewable.io -->

Konstantin8105 commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks broke.

---

*[main.go, line 28 at r2](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KvteT07u8vEJ2MDIb6s:-Kvzg-wfjZo9pv6k5gqJ:b-2mrct1) ([raw file](https://github.com/elliotchance/c2go/blob/97383239879874a4b1cbed2d0098911930cd122d/main.go#L28)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

I understand you want to maintain this package yourself, and you still can. However, this needs to be built in to `c2go` to be properly versioned with it.
</blockquote></details>

Yes, you are right. 
For solving that - we can use dependency management tool, for example https://github.com/golang/dep (In my point, it is ready to use), and we fixed all version of external library.
If you want, then I can do it, but in according to Internet discussions - programmers divided by 2 group: 
1) Include folder `vendor` inside git repository.
2) Do not include folder `vendor` inside git.
My opinion is second.
Please clarify your point of view.

---

*[main.go, line 220 at r2](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KvtfmieeoTmxo0KSmvt:-Kvzeg0KQFGI1QoiyhLM:b-896fix) ([raw file](https://github.com/elliotchance/c2go/blob/97383239879874a4b1cbed2d0098911930cd122d/main.go#L220)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

You need to print out the actual error here so people can report it as a bug:

```go
fmt.Fprintf(os.Stderr, "warning: cannot removing unused entities: %s", err.Error())
```
</blockquote></details>

Done.

---

*[main.go, line 233 at r2](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KvthNoG7IbzxZ1QDqaA:-KvzesDdphL2IA5LLEfz:bqmjwcz) ([raw file](https://github.com/elliotchance/c2go/blob/97383239879874a4b1cbed2d0098911930cd122d/main.go#L233)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

Please be more specific about what it being removed. Is it only constants and functions?
</blockquote></details>

Added.

---

*[main_test.go, line 317 at r2](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kvti1QEPBWXP-ja94bx:-Kvzfr10V2XLzpBvJRYy:b-896fix) ([raw file](https://github.com/elliotchance/c2go/blob/97383239879874a4b1cbed2d0098911930cd122d/main_test.go#L317)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

Please use https://golang.org/pkg/io/ioutil/#TempFile instead.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/elliotchance/c2go/235)*
<!-- Sent from Reviewable.io -->

elliotchance commented:
Reviewed 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks broke.

---

*[main.go, line 28 at r2](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KvteT07u8vEJ2MDIb6s:-Kw2VFh8KuiKVNdPkcIZ:b-cle119) ([raw file](https://github.com/elliotchance/c2go/blob/97383239879874a4b1cbed2d0098911930cd122d/main.go#L28)):*
<details><summary><i>Previously, Konstantin8105 (Konstantin) wrote…</i></summary><blockquote>

Yes, you are right. 
For solving that - we can use dependency management tool, for example https://github.com/golang/dep (In my point, it is ready to use), and we fixed all version of external library.
If you want, then I can do it, but in according to Internet discussions - programmers divided by 2 group: 
1) Include folder `vendor` inside git repository.
2) Do not include folder `vendor` inside git.
My opinion is second.
Please clarify your point of view.
</blockquote></details>

c2go already has some dependencies, but only for the testing code. This is OK as c2go will still be able to be built even if those tools break compatibility and the risk of removing or updating them is minimal because they are stable packages. However, I'm not ready to add production dependencies that are not stable.

You are free to add the features directly to c2go, but until a package is quite stable I can't have it as a production dependency. For the time being you could manage the code separately and merge it in at appropriate times so that it can be tested with c2go.

---

*[main.go, line 131 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2XKG7cxqYjhD5A3pt:-Kw2XKG7cxqYjhD5A3pu:b-h1gp89) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/main.go#L131)):*
> ```Go
> 		fmt.Println("Start tanspiling ...")
> 	}
> 	// recover of inside `c2go` panics
> ```

Do you have examples of why it would panic? This would be useful documentation here.

---

*[main.go, line 143 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2XTxz0FxRWzeWQqNn:-Kw2XTxz0FxRWzeWQqNo:b-9id8gi) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/main.go#L143)):*
> ```Go
> 
> 	if args.verbose {
> 		fmt.Println("Check environment variables GOPATH...")
> ```

This is too verbose. If it is successful this message is not helpful, if it fails then this message is also not helpful.

---

*[main.go, line 151 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2XXO6sDq92wM3maTs:-Kw2XXO6sDq92wM3maTt:b18jm97) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/main.go#L151)):*
> ```Go
> 	// 1. Compile it first (checking for errors)
> 	if args.verbose {
> 		fmt.Println("Check input file is exist...")
> ```

This is too verbose. The input file must always exist so this message is not helpful.

---

*[main.go, line 269 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2YEFNWEZKh1h_3d4j:-Kw2YEFNWEZKh1h_3d4k:b-vdanw1) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/main.go#L269)):*
> ```Go
> 		// remove unused
> 		if args.verbose {
> 			fmt.Println("Remove unused constants, functions, variables, types and methods of unused types...")
> ```

"Remove" -> "Removing"

---

*[main_test.go, line 113 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2YbgU7JDKXnUYFGbm:-Kw2YbgU7JDKXnUYFGbn:bi9iuck) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/main_test.go#L113)):*
> ```Go
> 				outputFile:  subFolder + mainFileName,
> 				packageName: "main",
> 				// add verbose for avoid travis timeout
> ```

You will have to explain this. How does making the output verbose affect Travis? Also, it has to work on Travis with the default options (which is not verbose)

---

*[README.md, line 5 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2_5dOLR3qdY7ogYw4:-Kw2_5dOLR3qdY7ogYw5:bvvcimo) ([raw file](https://github.com/elliotchance/c2go/blob/efb920dbbee77cc0418b48d93a551b44d09c7289/README.md#L5)):*
> ```Markdown
> [![codecov](https://codecov.io/gh/elliotchance/c2go/branch/master/graph/badge.svg)](https://codecov.io/gh/elliotchance/c2go)
> [![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/elliotchance/c2go/master/LICENSE)
> [![Join the chat at https://gitter.im/c2goproject](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/c2goproject?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
> ```

Why did you change this?

---

*[travis/test.sh, line 40 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2ZSeAAAarjPF3pG6d:-Kw2ZSeAAAarjPF3pG6e:b-3o7aww) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/travis/test.sh#L40)):*
> ```Shell
> echo "Run: go test"
> rm -f $OUTFILE
> travis 20 go list -f 'go test -v -tags integration -race -covermode atomic -coverprofile {{.Name}}.coverprofile -coverpkg $PKGS_DELIM {{.ImportPath}}' $PKGS | xargs -I{} bash -c "{} >> $OUTFILE"
> ```

Remove the `travis 20`. It looks like the build is only taking a few minutes.

---

*[travis/test.sh, line 43 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2ZahUlD04e0fEr1pK:-Kw2ZahUlD04e0fEr1pL:b-pro3b1) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/travis/test.sh#L43)):*
> ```Shell
> 
> # Merge coverage profiles.
> echo "Run: cover profile"
> ```

Why did you change all the code coverage?

---

*[travis/test.sh, line 78 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2Zl1Ryhz4GHra_2_h:-Kw2Zl1Ryhz4GHra_2_i:biuzxtt) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/travis/test.sh#L78)):*
> ```Shell
> export C2GO=$C2GO_DIR/c2go
> 
> echo "Run: transpile prime.c"
> ```

"Run: c2go transpile prime.c"

---

*[travis/test.sh, line 96 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2ZrNeDEaVMQyci0Fh:-Kw2ZrNeDEaVMQyci0Fi:bdojqii) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/travis/test.sh#L96)):*
> ```Shell
> # Add flag `-keep-unused` because linter have too many warning and
> # removing unused elements is not provided.
> ./c2go transpile -V -keep-unused /tmp/sqlite-amalgamation-3190300/shell.c
> ```

Does it take too long with `-keep-unused` disabled?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/elliotchance/c2go/235)*
<!-- Sent from Reviewable.io -->

Konstantin8105 commented:
Review status: 5 of 6 files reviewed at latest revision, 11 unresolved discussions, some commit checks broke.

---

*[main.go, line 28 at r2](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KvteT07u8vEJ2MDIb6s:-Kw3ultpSdDBNE1MCTnJ:bdqnaoy) ([raw file](https://github.com/elliotchance/c2go/blob/97383239879874a4b1cbed2d0098911930cd122d/main.go#L28)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

c2go already has some dependencies, but only for the testing code. This is OK as c2go will still be able to be built even if those tools break compatibility and the risk of removing or updating them is minimal because they are stable packages. However, I'm not ready to add production dependencies that are not stable.

You are free to add the features directly to c2go, but until a package is quite stable I can't have it as a production dependency. For the time being you could manage the code separately and merge it in at appropriate times so that it can be tested with c2go.
</blockquote></details>

Now, that code is outside of c2go(see https://github.com/Konstantin8105/Cleaner), I have a 83% coverage with tests generated in according to github.com/elliotchance/c2go/tests. I hope this is Ok for you at "stable" point of view.
I can move source to c2go, for example in some new folder `cleaner`.

---

*[main.go, line 131 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2XKG7cxqYjhD5A3pt:-Kw3wzswn6UXAPqb_hOS:b-1l8n8x) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/main.go#L131)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

Do you have examples of why it would panic? This would be useful documentation here.
</blockquote></details>

At the begin, I add because inside `main_test.go` we have test `TestStartPreprocess` , but on my laptop - it is panic-fail here
https://github.com/elliotchance/c2go/blob/efb920dbbee77cc0418b48d93a551b44d09c7289/main.go#L235 . After that I add recover and change panic to`fmt.Errorf`.
And No, I don`t have a example C code with `panic`, only panics inside https://github.com/elliotchance/c2go/search?utf8=%E2%9C%93&q=panic&type=

If it not Ok for you, then I will remove.

---

*[main.go, line 143 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2XTxz0FxRWzeWQqNn:-Kw3ypM7IRAWna1s8VPk:bzbi4r9) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/main.go#L143)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

This is too verbose. If it is successful this message is not helpful, if it fails then this message is also not helpful.
</blockquote></details>

Ok, remove.

---

*[main.go, line 151 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2XXO6sDq92wM3maTs:-Kw3ysucaQKkWKtRCw_C:bzbi4r9) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/main.go#L151)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

This is too verbose. The input file must always exist so this message is not helpful.
</blockquote></details>

Ok, remove.

---

*[main.go, line 269 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2YEFNWEZKh1h_3d4j:-Kw3yzhLmFm1-x30STvE:b-896fix) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/main.go#L269)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

"Remove" -> "Removing"
</blockquote></details>

Done.

---

*[main_test.go, line 113 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2YbgU7JDKXnUYFGbm:-Kw3znWZQeu9WLi1yfnM:by7uavt) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/main_test.go#L113)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

You will have to explain this. How does making the output verbose affect Travis? Also, it has to work on Travis with the default options (which is not verbose)
</blockquote></details>

The main problem : https://travis-ci.org/elliotchance/c2go/jobs/285526152 - travis terminated by timeout.
The reason (like I understood): https://github.com/elliotchance/c2go/blob/efb920dbbee77cc0418b48d93a551b44d09c7289/travis/test.sh#L37 . After add cleaner in c2go, that command execute is too long.
```
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
```
I try to solve that by solution:
1. expode the command on part with `echo` intermediant output
2. using command `travis_wait`
But both solution are fail on my hand.
** I need your help **

Also, I little bit don't undestood about verbose, if you compare next too travis output:
My travis logs with that `verbose: true,`: https://travis-ci.org/elliotchance/c2go/jobs/285603627
Your success tavis log : https://travis-ci.org/elliotchance/c2go/jobs/277055944
You will see many output in my logs.
Also, I add next line for correctness verbose https://github.com/Konstantin8105/c2go/blob/2c78a37c1ed80f18bfc920a78f3e41b98c7bad2b/main.go#L361

---

*[README.md, line 5 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2_5dOLR3qdY7ogYw4:-Kw41kKrl2rtXLixFwTk:btlyg40) ([raw file](https://github.com/elliotchance/c2go/blob/efb920dbbee77cc0418b48d93a551b44d09c7289/README.md#L5)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

Why did you change this?
</blockquote></details>

Markdown in Vim is fail on that part.
Excuse me, I forgot change back.

---

*[travis/test.sh, line 40 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2ZSeAAAarjPF3pG6d:-Kw42FuVFFBvPdWTexd5:b1wjyj0) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/travis/test.sh#L40)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

Remove the `travis 20`. It looks like the build is only taking a few minutes.
</blockquote></details>

Return back to your version.

---

*[travis/test.sh, line 43 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2ZahUlD04e0fEr1pK:-Kw42_v27gcojSLHvptM:b1wjyj0) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/travis/test.sh#L43)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

Why did you change all the code coverage?
</blockquote></details>

Return back to your version.

---

*[travis/test.sh, line 78 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2Zl1Ryhz4GHra_2_h:-Kw42dtJNnT2dj1mC-BL:blj75fd) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/travis/test.sh#L78)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

"Run: c2go transpile prime.c"
</blockquote></details>

Added at the on your design of script.

---

*[travis/test.sh, line 96 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2ZrNeDEaVMQyci0Fh:-Kw43aJ4AGoZdOnp9MZd:bd7x38y) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/travis/test.sh#L96)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

Does it take too long with `-keep-unused` disabled?
</blockquote></details>

It is added because linter of `cleaner` found too many errors in according to `golint` and `govet`. Now, design of `cleaner` is - "If we have golint or/and govet problem, then don`t clean the code". If you I can modify to design "Clean code also in spite of golint, govet".
Also, I check only by hand and sqlite haven`t unused entities.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/elliotchance/c2go/235)*
<!-- Sent from Reviewable.io -->

Konstantin8105 commented:
Review status: 3 of 6 files reviewed at latest revision, 11 unresolved discussions, some commit checks broke.

---

*[main_test.go, line 113 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2YbgU7JDKXnUYFGbm:-Kw6s6JZARuRKcWyROoB:b-nvuvo5) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/main_test.go#L113)):*
<details><summary><i>Previously, Konstantin8105 (Konstantin) wrote…</i></summary><blockquote>

The main problem : https://travis-ci.org/elliotchance/c2go/jobs/285526152 - travis terminated by timeout.
The reason (like I understood): https://github.com/elliotchance/c2go/blob/efb920dbbee77cc0418b48d93a551b44d09c7289/travis/test.sh#L37 . After add cleaner in c2go, that command execute is too long.
```
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
```
I try to solve that by solution:
1. expode the command on part with `echo` intermediant output
2. using command `travis_wait`
But both solution are fail on my hand.
** I need your help **

Also, I little bit don't undestood about verbose, if you compare next too travis output:
My travis logs with that `verbose: true,`: https://travis-ci.org/elliotchance/c2go/jobs/285603627
Your success tavis log : https://travis-ci.org/elliotchance/c2go/jobs/277055944
You will see many output in my logs.
Also, I add next line for correctness verbose https://github.com/Konstantin8105/c2go/blob/2c78a37c1ed80f18bfc920a78f3e41b98c7bad2b/main.go#L361

</blockquote></details>

I find new solution- we can execute bash scripts from folder 'travis' not directly, but we can execute small go program with 2gorotines : one for run script and one for print time every 30 seconds. In my point of it easy and solve that problem forever.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/elliotchance/c2go/235)*
<!-- Sent from Reviewable.io -->

elliotchance commented:
Reviewed 2 of 3 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[main.go, line 28 at r2](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KvteT07u8vEJ2MDIb6s:-KwCaz17n_tOBG9K8tlO:b-f8xktr) ([raw file](https://github.com/elliotchance/c2go/blob/97383239879874a4b1cbed2d0098911930cd122d/main.go#L28)):*
<details><summary><i>Previously, Konstantin8105 (Konstantin) wrote…</i></summary><blockquote>

Now, that code is outside of c2go(see https://github.com/Konstantin8105/Cleaner), I have a 83% coverage with tests generated in according to github.com/elliotchance/c2go/tests. I hope this is Ok for you at "stable" point of view.
I can move source to c2go, for example in some new folder `cleaner`.
</blockquote></details>

A high code coverage does not mean that a piece of software is stable, in fact it doesn't even mean the software works.

A stable piece of software has been around for a while, used by a lot of people and uncovered/fixed bugs (battle tested. This leads to an API that is very unlikely to change.

On the other hand new packages (even those that are bug free) are much more likely to change independently of the things that depend on them because they are new and haven't discovered a suitable way to make everyone happy.

`c2go/cleaner` will do for now. If you also wish to maintain it as a separate package you will be in charge to bringing the new features and bug fixes through a PR. This means that the code from now and in the future will have to be the same standard as any other PR against c2go.

Likewise, others can fix bugs. You will be able to these cherry-pick back to your repository as you see necessary.

---

*[main.go, line 131 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2XKG7cxqYjhD5A3pt:-KwCeE2DrBRN2J643_bf:b8dt0wc) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/main.go#L131)):*
<details><summary><i>Previously, Konstantin8105 (Konstantin) wrote…</i></summary><blockquote>

At the begin, I add because inside `main_test.go` we have test `TestStartPreprocess` , but on my laptop - it is panic-fail here
https://github.com/elliotchance/c2go/blob/efb920dbbee77cc0418b48d93a551b44d09c7289/main.go#L235 . After that I add recover and change panic to`fmt.Errorf`.
And No, I don`t have a example C code with `panic`, only panics inside https://github.com/elliotchance/c2go/search?utf8=%E2%9C%93&q=panic&type=

If it not Ok for you, then I will remove.
</blockquote></details>

Please remove as we want to remove cases where panics happen and not encourage them as suitable.

---

*[main_test.go, line 113 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2YbgU7JDKXnUYFGbm:-KwCefyhy3ID1thPF0M7:boibv2d) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/main_test.go#L113)):*
<details><summary><i>Previously, Konstantin8105 (Konstantin) wrote…</i></summary><blockquote>

I find new solution- we can execute bash scripts from folder 'travis' not directly, but we can execute small go program with 2gorotines : one for run script and one for print time every 30 seconds. In my point of it easy and solve that problem forever.
</blockquote></details>

These are not solutions, they are just trying to avoid the problem. The real problem is that the execution takes too long. Can you show me how much time difference it makes when you run the test suite on your machine with and without the `-keep-used` option?

---

*[main_test.go, line 124 at r5](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwCfJGxLv1x4dvDc94-:-KwCfJGxLv1x4dvDc940:b-wbfm2v) ([raw file](https://github.com/elliotchance/c2go/blob/0ca8ec77e76f94d1ee889f19ffdffe6c6cd6344d/main_test.go#L124)):*
> ```Go
> 			if file == "tests/stdlib.c" {
> 				programArgs.keepUnused = false
> 				programArgs.verbose = true
> ```

I'm putting a marker here because this may need to be changed based on my previous comment for timings.

---

*[travis/test.sh, line 96 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2ZrNeDEaVMQyci0Fh:-KwCfZFWR4Sb0msp_WCd:b-6u6rqn) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/travis/test.sh#L96)):*
<details><summary><i>Previously, Konstantin8105 (Konstantin) wrote…</i></summary><blockquote>

It is added because linter of `cleaner` found too many errors in according to `golint` and `govet`. Now, design of `cleaner` is - "If we have golint or/and govet problem, then don`t clean the code". If you I can modify to design "Clean code also in spite of golint, govet".
Also, I check only by hand and sqlite haven`t unused entities.
</blockquote></details>

I don't understand. Are you saying that c2go should perform differently if `govet` or `golint` are also installed?

Also, please remove the verbose option.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/elliotchance/c2go/235)*
<!-- Sent from Reviewable.io -->

Konstantin8105 commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[main.go, line 28 at r2](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KvteT07u8vEJ2MDIb6s:-KwR0LfZTgl-5XAucNa_:bm4nma3) ([raw file](https://github.com/elliotchance/c2go/blob/97383239879874a4b1cbed2d0098911930cd122d/main.go#L28)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

A high code coverage does not mean that a piece of software is stable, in fact it doesn't even mean the software works.

A stable piece of software has been around for a while, used by a lot of people and uncovered/fixed bugs (battle tested. This leads to an API that is very unlikely to change.

On the other hand new packages (even those that are bug free) are much more likely to change independently of the things that depend on them because they are new and haven't discovered a suitable way to make everyone happy.

`c2go/cleaner` will do for now. If you also wish to maintain it as a separate package you will be in charge to bringing the new features and bug fixes through a PR. This means that the code from now and in the future will have to be the same standard as any other PR against c2go.

Likewise, others can fix bugs. You will be able to these cherry-pick back to your repository as you see necessary.
</blockquote></details>

One more think - my library in present design based on next external open source libraries:
```golang
	"honnef.co/go/tools/lint/lintutil"
	"honnef.co/go/tools/unused"
```
that library is create for Go tool `unused` and that tool is included in `gometalinter`.

I will move to code your project and add more comments for clarification of "How it is work".

---

*[main.go, line 131 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2XKG7cxqYjhD5A3pt:-KwR41cPErjSj7VklXHI:btsrb8y) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/main.go#L131)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

Please remove as we want to remove cases where panics happen and not encourage them as suitable.
</blockquote></details>

Ok. I will do it.

---

*[main_test.go, line 113 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2YbgU7JDKXnUYFGbm:-KwR-WDnc3GIABq_k_jc:b5w4j9i) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/main_test.go#L113)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

These are not solutions, they are just trying to avoid the problem. The real problem is that the execution takes too long. Can you show me how much time difference it makes when you run the test suite on your machine with and without the `-keep-used` option?
</blockquote></details>

if we use `keep unused` for all test files:
```
ok  	github.com/elliotchance/c2go	347.765s
real 6m04s
user 6m53s
sys  0m31s
```

if we don't use:
```
ok  	github.com/elliotchance/c2go	529.343s
real  8m58s
user 14m14s
sys   0m41s
```

on both cases run `time go test -tags=integration ./...`

---

*[main_test.go, line 124 at r5](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwCfJGxLv1x4dvDc94-:-KwR-UdqRdjHo-aEE8o_:b32wtd) ([raw file](https://github.com/elliotchance/c2go/blob/0ca8ec77e76f94d1ee889f19ffdffe6c6cd6344d/main_test.go#L124)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

I'm putting a marker here because this may need to be changed based on my previous comment for timings.
</blockquote></details>

ok

---

*[travis/test.sh, line 96 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2ZrNeDEaVMQyci0Fh:-KwR5IcslR_WxZUo38ml:b-k2ctlc) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/travis/test.sh#L96)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

I don't understand. Are you saying that c2go should perform differently if `govet` or `golint` are also installed?

Also, please remove the verbose option.
</blockquote></details>

Ok, I will do it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/elliotchance/c2go/235)*
<!-- Sent from Reviewable.io -->

elliotchance commented:
Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[main.go, line 60 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwggPNmAytYCebSVaRp:-KwggPNmAytYCebSVaRq:b-eba8du) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/main.go#L60)):*
> ```Go
> 
> 	// Keep unused entities
> 	keepUnused bool
> ```

1. Please create a method called `DefaultProgramArgs()` that returns a `ProgramArgs` with every attribute set to a default.
2. Find any cases of `ProgramArgs{}` and replace with `DefaultProgramArgs()`.
3. Update the documentation for `ProgramArgs` with "Do not instantiate this directly. Instead use DefaultProgramArgs(); then modify any specific attributes."

---

*[main.go, line 215 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwghjnESarI3SLVrM6d:-KwghjnESarI3SLVrM6e:b-u1ojlr) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/main.go#L215)):*
> ```Go
> 	// build tree
> 	if args.verbose {
> 		fmt.Println("Build a tree...")
> ```

"Building tree..."

---

*[main.go, line 232 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwghpJqE-WMqhvHXQ_x:-KwghpJqE-WMqhvHXQ_y:bd1qmvs) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/main.go#L232)):*
> ```Go
> 	// transpile ast tree
> 	if args.verbose {
> 		fmt.Println("Transpile AST tree...")
> ```

"Transpiling tree..."

---

*[main.go, line 317 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwgiK6qsVzPlkgB7uQc:-KwgiK6qsVzPlkgB7uQd:bijo1oo) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/main.go#L317)):*
> ```Go
> 	}
> 
> 	args := ProgramArgs{ast: false}
> ```

Here is one place to replace with `DefaultProgramArgs()`.

---

*[main_test.go, line 113 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2YbgU7JDKXnUYFGbm:-KwgijUkAXyXA9kAuVTo:byy3ks6) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/main_test.go#L113)):*
<details><summary><i>Previously, Konstantin8105 (Konstantin) wrote…</i></summary><blockquote>

if we use `keep unused` for all test files:
```
ok  	github.com/elliotchance/c2go	347.765s
real 6m04s
user 6m53s
sys  0m31s
```

if we don't use:
```
ok  	github.com/elliotchance/c2go	529.343s
real  8m58s
user 14m14s
sys   0m41s
```

on both cases run `time go test -tags=integration ./...`
</blockquote></details>

Thanks for that information. I know this slow down the build, but it it better to test everything with unused entities removed. Set this to `keepUnused: false` and remove the comment about travis.

---

*[main_test.go, line 124 at r5](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwCfJGxLv1x4dvDc94-:-KwgiuO7jQSEjHPOy5r3:b-o1w4ij) ([raw file](https://github.com/elliotchance/c2go/blob/0ca8ec77e76f94d1ee889f19ffdffe6c6cd6344d/main_test.go#L124)):*
<details><summary><i>Previously, Konstantin8105 (Konstantin) wrote…</i></summary><blockquote>

ok
</blockquote></details>

Remove this whole `if` statement. Since all files will be tested with this option. If this is a problem we could add a test CLI option to enable/disbale this later.

---

*[cleaner/cleaner.go, line 23 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kwgjz2O7lrfZSVvTT8J:-Kwgjz2PyHwvtQPLNoBb:b-i8rhr2) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/cleaner/cleaner.go#L23)):*
> ```Go
> 
> const (
> 	unusedConstantans un = iota
> ```

`unusedConstants`

---

*[cleaner/cleaner.go, line 82 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwgkHBfQZLuzUzklAEh:-KwgkHBfQZLuzUzklAEi:bzbe37d) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/cleaner/cleaner.go#L82)):*
> ```Go
> 	var mode unused.CheckMode
> 	mode |= unused.CheckConstants
> 	//mode |= unused.CheckFields
> ```

What is this for?

---

*[cleaner/cleaner.go, line 165 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwgkX9qlMxzbfahLP8C:-KwgkX9qlMxzbfahLP8D:b3ge8sf) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/cleaner/cleaner.go#L165)):*
> ```Go
> 					gen, ok := tree.Decls[i].(*ast.GenDecl)
> 					if !ok || gen == (*ast.GenDecl)(nil) || gen.Tok != token.CONST {
> 						goto nextConstDecl
> ```

Would a `continue` here do the same thing?

---

*[cleaner/cleaner.go, line 185 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kwgkj0xTDw1WNDl6JGD:-Kwgkj0xTDw1WNDl6JGE:b-p51khr) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/cleaner/cleaner.go#L185)):*
> ```Go
> 					gen, ok := tree.Decls[i].(*ast.FuncDecl)
> 					if !ok || gen == (*ast.FuncDecl)(nil) {
> 						goto nextFuncDecl
> ```

`continue`?

---

*[cleaner/cleaner.go, line 202 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwgkpxDl1QqifeaN9hT:-KwgkpxDl1QqifeaN9hU:b-p51khr) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/cleaner/cleaner.go#L202)):*
> ```Go
> 					gen, ok := tree.Decls[i].(*ast.GenDecl)
> 					if !ok || gen == (*ast.GenDecl)(nil) || gen.Tok != token.TYPE {
> 						goto nextTypeDecl
> ```

`continue`?

---

*[cleaner/cleaner.go, line 220 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwgkrXSltWTy6fiXev9:-KwgkrXSltWTy6fiXevA:b-p51khr) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/cleaner/cleaner.go#L220)):*
> ```Go
> 					gen, ok := tree.Decls[i].(*ast.GenDecl)
> 					if !ok || gen == (*ast.GenDecl)(nil) || gen.Tok != token.VAR {
> 						goto nextVarDecl
> ```

`continue`?

---

*[cleaner/cleaner.go, line 318 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwglBN8OwYbLaRkC9gQ:-KwglBN8OwYbLaRkC9gR:be08rq5) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/cleaner/cleaner.go#L318)):*
> ```Go
> 
> // newTempFile - returns temp file
> func newTempFile(dir, prefix string) (tmpFile *os.File, err error) {
> ```

I think Go already has a function for this?

---

*[travis/test.sh, line 76 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwglGljQVzR9ZZeGRlY:-KwglGljQVzR9ZZeGRlZ:b-xy8jpc) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/travis/test.sh#L76)):*
> ```Shell
> unzip /tmp/$SQLITE3_FILE.zip -d /tmp
> 
> # Transpile the SQLite3 files.
> ```

How much longer does the build take when `-keep-unused` is removed?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/elliotchance/c2go/235)*
<!-- Sent from Reviewable.io -->

Konstantin8105 commented:
Review status: 6 of 7 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[main.go, line 60 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwggPNmAytYCebSVaRp:-Kwq6a3_twaXEjwA2IbW:b-896fix) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/main.go#L60)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

1. Please create a method called `DefaultProgramArgs()` that returns a `ProgramArgs` with every attribute set to a default.
2. Find any cases of `ProgramArgs{}` and replace with `DefaultProgramArgs()`.
3. Update the documentation for `ProgramArgs` with "Do not instantiate this directly. Instead use DefaultProgramArgs(); then modify any specific attributes."
</blockquote></details>

Done.

---

*[main.go, line 215 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwghjnESarI3SLVrM6d:-KwpyZiaeqqx5OdLie6v:bb7f2mf) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/main.go#L215)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

"Building tree..."
</blockquote></details>

done

---

*[main.go, line 232 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwghpJqE-WMqhvHXQ_x:-Kwq2ATrVbcfMXKRSISL:b-896fix) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/main.go#L232)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

"Transpiling tree..."
</blockquote></details>

Done.

---

*[main.go, line 317 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwgiK6qsVzPlkgB7uQc:-Kwq6eda9rMeVJbIhpyq:b-896fix) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/main.go#L317)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

Here is one place to replace with `DefaultProgramArgs()`.
</blockquote></details>

Done.

---

*[main_test.go, line 113 at r3](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kw2YbgU7JDKXnUYFGbm:-KwpyvuG79JRQ8_-lR7L:b-woo63l) ([raw file](https://github.com/elliotchance/c2go/blob/c6fffc137434dbada39c4fd09f75dd72efcacdd0/main_test.go#L113)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

Thanks for that information. I know this slow down the build, but it it better to test everything with unused entities removed. Set this to `keepUnused: false` and remove the comment about travis.
</blockquote></details>

Done. I will try to rebase to your master branch for using Regexp optimization.

---

*[main_test.go, line 124 at r5](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwCfJGxLv1x4dvDc94-:-KwpzCkKzA2WqbniXI-j:ba3dfob) ([raw file](https://github.com/elliotchance/c2go/blob/0ca8ec77e76f94d1ee889f19ffdffe6c6cd6344d/main_test.go#L124)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

Remove this whole `if` statement. Since all files will be tested with this option. If this is a problem we could add a test CLI option to enable/disbale this later.
</blockquote></details>

Removed

---

*[cleaner/cleaner.go, line 23 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kwgjz2O7lrfZSVvTT8J:-KwpzQ6__1zVswiifGJP:b-896fix) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/cleaner/cleaner.go#L23)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

`unusedConstants`
</blockquote></details>

Done.

---

*[cleaner/cleaner.go, line 82 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwgkHBfQZLuzUzklAEh:-KwpzXaxq1L8FqjtRrrT:b-fzna80) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/cleaner/cleaner.go#L82)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

What is this for?
</blockquote></details>

Yes, If we add this option, then we can found the unused fields of structs.
Now, I don't not sure to use that.
Now, I remove that mode.

---

*[cleaner/cleaner.go, line 165 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwgkX9qlMxzbfahLP8C:-Kwq1qiJSDfVLP1oIVU3:b-896fix) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/cleaner/cleaner.go#L165)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

Would a `continue` here do the same thing?
</blockquote></details>

Done.

---

*[cleaner/cleaner.go, line 185 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kwgkj0xTDw1WNDl6JGD:-Kwq1o-8mxuKFwOT1jSF:b-7rl0jd) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/cleaner/cleaner.go#L185)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

`continue`?
</blockquote></details>

done.

---

*[cleaner/cleaner.go, line 202 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwgkpxDl1QqifeaN9hT:-Kwq0_mg3udoS9HdDwB6:b-qq73hf) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/cleaner/cleaner.go#L202)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

`continue`?
</blockquote></details>

Yes, you are right.
replace to `continue`

---

*[cleaner/cleaner.go, line 220 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwgkrXSltWTy6fiXev9:-Kwq0gxUF-F-EfjLTXvo:b-41w4tv) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/cleaner/cleaner.go#L220)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

`continue`?
</blockquote></details>

Replace to `continue`

---

*[cleaner/cleaner.go, line 318 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwglBN8OwYbLaRkC9gQ:-Kwq1MDuqA0ijtEdT_KJ:bb7f2mf) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/cleaner/cleaner.go#L318)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

I think Go already has a function for this?
</blockquote></details>

done

---

*[travis/test.sh, line 76 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwglGljQVzR9ZZeGRlY:-Kwq1RLydswYAQ6vuZRS:bv7dbsd) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/travis/test.sh#L76)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

How much longer does the build take when `-keep-unused` is removed?
</blockquote></details>

Now, I remove and we will see travis time of checking.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/elliotchance/c2go/235)*
<!-- Sent from Reviewable.io -->

elliotchance commented:
Reviewed 4 of 4 files at r7.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[cleaner/cleaner.go, line 23 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kwgjz2O7lrfZSVvTT8J:-Kwqpt7KLlw4PeJt2sF3:buwj8u7) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/cleaner/cleaner.go#L23)):*
<details><summary><i>Previously, Konstantin8105 (Konstantin) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Almost, just missing a `t`: "unusedConstants"

---

*[cleaner/cleaner.go, line 82 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwgkHBfQZLuzUzklAEh:-Kwqq4phP2Zt6ohNovLt:b-gmepwu) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/cleaner/cleaner.go#L82)):*
<details><summary><i>Previously, Konstantin8105 (Konstantin) wrote…</i></summary><blockquote>

Yes, If we add this option, then we can found the unused fields of structs.
Now, I don't not sure to use that.
Now, I remove that mode.
</blockquote></details>

You are right, we definitely do not what to do that. You can remove the line.

---

*[cleaner/cleaner.go, line 165 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwgkX9qlMxzbfahLP8C:-KwqrFEh-fvZwCeP2dre:btj7xad) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/cleaner/cleaner.go#L165)):*
<details><summary><i>Previously, Konstantin8105 (Konstantin) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

This one has not been replaced?

---

*[travis/test.sh, line 76 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwglGljQVzR9ZZeGRlY:-Kwqqif6jW7zgKM4uts5:be2ciph) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/travis/test.sh#L76)):*
<details><summary><i>Previously, Konstantin8105 (Konstantin) wrote…</i></summary><blockquote>

Now, I remove and we will see travis time of checking.
</blockquote></details>

It looks like the build it less than 10 minute. This is good news - let's keep it this way.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/elliotchance/c2go/235)*
<!-- Sent from Reviewable.io -->

Konstantin8105 commented:
Reviewed 1 of 3 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[cleaner/cleaner.go, line 23 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kwgjz2O7lrfZSVvTT8J:-Kwsmpzl_hfBEUBZ7Rmi:bal91si) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/cleaner/cleaner.go#L23)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

Almost, just missing a `t`: "unusedConstants"
</blockquote></details>

Excuse me.

---

*[cleaner/cleaner.go, line 82 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwgkHBfQZLuzUzklAEh:-KwsmvYzdO8lAfZqFnyk:b-896fix) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/cleaner/cleaner.go#L82)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

You are right, we definitely do not what to do that. You can remove the line.
</blockquote></details>

Done.

---

*[cleaner/cleaner.go, line 165 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwgkX9qlMxzbfahLP8C:-Kwsn7pI7etyeelnXOBi:bb6un1j) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/cleaner/cleaner.go#L165)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

This one has not been replaced?
</blockquote></details>

Done

---

*[travis/test.sh, line 76 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwglGljQVzR9ZZeGRlY:-KwsnLCs_IaQm04sVm3S:b-1l9ug3) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/travis/test.sh#L76)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

It looks like the build it less than 10 minute. This is good news - let's keep it this way.
</blockquote></details>

In according to travis results of that PR after performance optimization - all is Ok without travis timeout break.
But we still near to 10 min limit - https://travis-ci.org/elliotchance/c2go/builds/290187036?utm_source=github_status&utm_medium=notification.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/elliotchance/c2go/235)*
<!-- Sent from Reviewable.io -->

Konstantin8105 commented:
@elliotchance , We have to close that PR, because - see #232

Konstantin8105 commented:
Review status: 5 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---

*[cleaner/cleaner.go, line 23 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-Kwgjz2O7lrfZSVvTT8J:-KwvbKuUzCY1jlyNQejg:b4uucet) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/cleaner/cleaner.go#L23)):*
<details><summary><i>Previously, Konstantin8105 (Konstantin) wrote…</i></summary><blockquote>

Excuse me.
</blockquote></details>

We have to close that PR, because - see #232
I will extract good thinks like (verbose print, misspel) to another PR.

---

*[cleaner/cleaner.go, line 82 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwgkHBfQZLuzUzklAEh:-KwvbNP9wCChAgTugsA0:b4uucet) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/cleaner/cleaner.go#L82)):*
<details><summary><i>Previously, Konstantin8105 (Konstantin) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

We have to close that PR, because - see #232
I will extract good thinks like (verbose print, misspel) to another PR.

---

*[cleaner/cleaner.go, line 165 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwgkX9qlMxzbfahLP8C:-KwvbQAqOGchKigcfK1Z:b4uucet) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/cleaner/cleaner.go#L165)):*
<details><summary><i>Previously, Konstantin8105 (Konstantin) wrote…</i></summary><blockquote>

Done
</blockquote></details>

We have to close that PR, because - see #232
I will extract good thinks like (verbose print, misspel) to another PR.

---

*[travis/test.sh, line 76 at r6](https://reviewable.io:443/reviews/elliotchance/c2go/235#-KwglGljQVzR9ZZeGRlY:-KwvbTOovmxdpTG0gE9J:b4uucet) ([raw file](https://github.com/elliotchance/c2go/blob/9ba2fdb414ab260949ac23f7ffb27554d41d5e9f/travis/test.sh#L76)):*
<details><summary><i>Previously, Konstantin8105 (Konstantin) wrote…</i></summary><blockquote>

In according to travis results of that PR after performance optimization - all is Ok without travis timeout break.
But we still near to 10 min limit - https://travis-ci.org/elliotchance/c2go/builds/290187036?utm_source=github_status&utm_medium=notification.
</blockquote></details>

We have to close that PR, because - see #232
I will extract good thinks like (verbose print, misspel) to another PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/elliotchance/c2go/235)*
<!-- Sent from Reviewable.io -->

elliotchance commented:
Reviewed 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks broke.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/elliotchance/c2go/235)*
<!-- Sent from Reviewable.io -->

Konstantin8105 commented:
Ready to merge

elliotchance commented:
Reviewed 2 of 4 files at r9, 1 of 1 files at r10.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/elliotchance/c2go/235)*
<!-- Sent from Reviewable.io -->
